docs/doc/60-contributing/03-rfcs/20220725-cache.md

```diff
@@ -0,0 +1,151 @@
+---
+title: Cache
+description: Cache for Databend Query
+---
+
+- RFC PR: [datafuselabs/databend#6799](https://github.com/datafuselabs/databend/pull/6799)
+- Tracking Issue: [datafuselabs/databend#6786](https://github.com/datafuselabs/databend/issues/6786)
+
+# Summary
+
+Add cache support for Databend Query so that our users can load hot data from our cache services instead of reading from
+backend storage services again.
+
+# Motivation
+
+Databend design is based on the separation of storage and computing:
+
+![](/img/rfc/20220725-cache/architecture.svg)
+
+As a result, databend will support different cloud storage services as its backend like AWS s3, azure blob, and google cloud
+storage. The advantage of this is that we can get very high throughput bandwidth, but the disadvantage is that the
+latency of our individual requests increases.
+
+Adding a caching layer will improve the overall latency and reduce the number of unnecessary requests, thus reducing
+costs.
+
+Databend used to have a local disk-based cache
+at [common-cache](https://github.com/datafuselabs/databend/tree/cd5124c4fbf53e8992f820c7b0a0fbc6442876e6/common/cache).
+This implementation use databend-query's local path to cache blocks.
+
+# Guide-level explanation
+
+After introducing cache, the databend storage layer will be composed as follows:
+
+![](/img/rfc/20220725-cache/cache.png)
+
+The query will maintain three OpenDAL operators:
+
+- Persist Operator: Read and write persist data from/into storage services **directly**.
+- Cache Operator: Read and write data via transparent caching.
+- Temporary Operator: Read and write temporary data from/into cache services.
+
+Storage services and cache services we mentioned here are both OpenDAL-supported services. The difference is cache services'
+backends could have their GC or background auto eviction logic, which means cache services is non-persist, a.k.a,
+volatile.
+
+The query should never try to write persisted data to `Cache Operator` or `Temporary Operator`.
+
+## For End Users
+
+Users can specify different cache services for caching or storing temporary data. For example, they can cache data into
+local fs (the same as current behavior):
+
+```toml
+[cache]
+type = "fs"
+
+[cache.fs]
+data_path = "/var/cache/databend/"
+```
+
+Or they can cache data into Redis:
+
+```toml
+[cache]
+type = "redis"
+
+[cache.redis]
+endpoint = "192.168.1.2"
+password = "PASSWORD"
+```
+
+Temporary data could be shared crossing the whole cluster, so users must use distributed cache services instead of
+local fs or memory:
+
+```toml
+[temporary]
+type = "opencache"
```

Comment:
opencache is not for temporary data, I think it's a kind of cache server such as Redis and can be set the TTL?

Reply:
> opencache is not for temporary data, I think it's a kind of cache server such as Redis and can be set the TTL?

Right... As I commented in: https://github.com/datafuselabs/databend/issues/4521#issuecomment-1193602317

- Temporary store has to be durable and persistent. The eviction policy is least-recently-added.
- Block cache is as you described in RFC, **volatile**. And the eviction policy is more like to be least-recently-used.

But this is not a big difference in implementation: 
- Temporary store requires `fsync` on write and adopts least-recently-added eviction policy.
- Block cache does not require `fsync` on write and adopts a least-recently-used eviction policy.

Another issue with the temporary store is that the hash method on the client side must be consistent on every query node. Otherwise, a query node can not see the temp data written by another query node with a different hash configuration.

Reply:
Thanks for the explanation, I will add it to RFC.

Reply:
Addressed, PTAL.